sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+      }
+      /*
+       * Validates key/value class with InputFormat's parameterized type.
+       */
+      if (inputClass != null) {
+        return validateClassesEquality(inputClass, property);
+      }
+      return true;
+    }
+
+    /**
+     * Returns true if first record's key/value encodes and decodes successfully. Also sets
+     * expectedKeyClass/expectedValueClass if cloning of key/value fails.
+     */
+    private <T> boolean validateClassUsingCoder(String property, Coder<T> coder)
+        throws IOException, InterruptedException {
+      final RecordReader<?, ?> reader;
+      reader = fetchFirstRecord();
+      if (property.contains("key")) {
+        boolean isKeyClassValide = checkEncodingAndDecoding(coder, (T) reader.getCurrentKey());
+        if (!isKeyClassValide) {
+          this.setExpectedKeyClass(reader.getCurrentKey().getClass());
+        }
+        return isKeyClassValide;
+      } else {
+        boolean isValueClassValide = checkEncodingAndDecoding(coder, (T) reader.getCurrentValue());
+        if (!isValueClassValide) {
+          this.setExpectedValueClass(reader.getCurrentValue().getClass());
+        }
+        return isValueClassValide;
+      }
+    }
+
+    /**
+     * Returns true if the class set in a given property by the pipeline user is type of given
+     * expectedClass.
+     */
+    private boolean validateClassesEquality(Class<?> expectedClass, String property) {
+      Class<?> actualClass = conf.getHadoopConfiguration().getClass(property, Object.class);
+      if (!actualClass.isAssignableFrom(expectedClass)) {
+        if (property.contains("key")) {
+          this.setExpectedKeyClass(expectedClass);
+        } else {
+          this.setExpectedValueClass(expectedClass);
+        }
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Returns false if exception occurs while encoding and decoding the given value with the given
+     * coder.
+     */
+    private <T> boolean checkEncodingAndDecoding(Coder<T> coder, T value) {
+      try {
+        CoderUtils.clone(coder, value);
+      } catch (Exception e) {
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Returns parameterized type of the InputFormat class.
+     */
+    private ParameterizedType determineGenericType() {
+      Class<?> inputFormatClass = inputFormatObj.getClass();
+      Type genericSuperclass = null;
+      for (;;) {
+        genericSuperclass = inputFormatClass.getGenericSuperclass();
+        if (genericSuperclass instanceof ParameterizedType)
+          break;
+        inputFormatClass = inputFormatClass.getSuperclass();
+      }
+      return (ParameterizedType) genericSuperclass;
+    }
+
+    /**
+     * Returns RecordReader object of the first split to read first record for validating key/value
+     * classes.
+     */
+    private RecordReader fetchFirstRecord() throws IOException, InterruptedException {
```

Comment:
this name is a bit deceptive - it claims it returns the first record, but it's actually returning the first record reader.